tools/setup-envtest/remote/client.go

```diff
@@ -0,0 +1,217 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2021 The Kubernetes Authors
+
+package remote
+
+import (
+	"context"
+	"crypto/md5"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+	"regexp"
+	"sort"
+
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/tools/setup-envtest/versions"
+)
+
+var (
+	// versionRERaw is the internals of VersionRE
+	versionRERaw = `(?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)-(?P<os>\w+)-(?P<arch>\w+)`
+	// VersionRE matches concrete version-platform strings.
+	VersionRE = regexp.MustCompile("^" + versionRERaw + "$")
+	// ArchiveRE matches concrete version-platform.tar.gz strings.
+	ArchiveRE = regexp.MustCompile(`^kubebuilder-tools-` + versionRERaw + `\.tar\.gz$`)
+)
+
+// objectList is the parts we need of the GCS "list-objects-in-bucket" endpoint.
+type objectList struct {
+	Items         []bucketObject `json:"items"`
+	NextPageToken string         `json:"nextPageToken"`
+}
+
+// bucketObject is the parts we need of the GCS object metadata.
+type bucketObject struct {
+	Name string `json:"name"`
+	Hash string `json:"md5Hash"`
+}
+
+// Client is a basic client for fetching versions of the envtest binary archives
+// from GCS.
+type Client struct {
+	// Bucket is the bucket to fetch from.
+	Bucket string
+
+	// Log allows us to log.
+	Log logr.Logger
+}
+
+// ListVersions lists all available tools versions in the given bucket, along
+// with supported os/arch combos and the corresponding hash.
+//
+// The results are sorted with newer versions first.
+func (c *Client) ListVersions(ctx context.Context) ([]versions.Set, error) {
+	loc := &url.URL{
```

Comment:
Can you not use the [google storage API](https://pkg.go.dev/cloud.google.com/go/storage) here?

Reply:
didn't want to pull in a whole bunch of deps for basically one API call, plus testing is a bit easier.

Reply:
I can switch it out if you want, but it's not really much code.

Reply:
Nah that's a fair point.